File: thys/BirdKMP/document/programs/KMP.ml
(* Bird's Morris-Pratt string matcher, with the `K' optimisation in ocaml
   Chapter 17, "Pearls of Functional Algorithm Design", 2010. *)

(*
Encoding the requisite laziness is fiddly:
 - as Wadler observes this style is `odd`
 - forced to use `Lazy.t` to share `root`
 - intuitively need to manually manage how demand is transferred
   through the computation.
*)

type 'a tree
  = Null
  | Node of 'a list * 'a ltree * 'a tree
and 'a ltree = 'a tree Lazy.t

let kmatches (type a) (eq: a -> a -> bool) (ws: a list) : a list -> int list =
     let ok (t: a ltree) : bool = match Lazy.force t with Node (vs, l, r) -> vs = []
  in let next (x: a) (t: a ltree) : a ltree =
           lazy (let t = Lazy.force t in match t with
             Null -> Null
           | Node ([], _l, _r) -> t
           | Node (v :: vs, l, _r) -> if eq x v then Lazy.force l else t)
  in let rec op (t: a ltree) (x: a) : a ltree =
           lazy (match Lazy.force t with
             Null -> Lazy.force root
           | Node (vvs, l, r) ->
             (match vvs with
                [] -> Lazy.force (op l x)
              | v :: vs -> if eq x v then r else Lazy.force (op l x)))
     and grep (l: a ltree) (vvs: a list) : a tree =
           match vvs with
             [] -> Node ([], l, Null)
           | v :: vs -> Node (vvs, next v l, grep (op l v) vs)
     and root : a ltree = lazy (grep (lazy Null) ws)
  in let step (nt: int * a ltree) (x: a) : int * a ltree =
           fst nt + 1, op (snd nt) x
  in let rec rheight (t: a tree) =
           match t with Null -> 0 | Node (_, _, r) -> 1 + rheight r
  in let rec driver (nt: int * a ltree) (xxs: a list) : int list =
           match xxs with
             [] -> if ok (snd nt) then [fst nt] else []
           | x :: xs -> let nt' : int * a ltree = step nt x
                    in if ok (snd nt) then fst nt :: driver nt' xs else driver nt' xs
  in driver (0, root)

;;

List.iter (Printf.printf "%d ") (kmatches (=) [] []) ;
List.iter (Printf.printf "%d ") (kmatches (=) [] [1;2;3]) ;
List.iter (Printf.printf "%d ") (kmatches (=) [1;2;1] []) ;
List.iter (Printf.printf "%d ") (kmatches (=) [1;2] [1;2;3]) ;
List.iter (Printf.printf "%d ") (kmatches (=) [1;2;3;1;2] [1;2;1;2;3;1;2;3;1;2]) ;

File: thys/BirdKMP/document/programs/KMP.pl
% -*- mode: prolog -*-
% Bird's Morris-Pratt string matcher
% Chapter 17, "Pearls of Functional Algorithm Design", 2010.
%   - adapted to use rational trees.
%   - with the `K' (next) optimisation
% Tested with SWI Prolog, which has good support for rational trees.

% root/2 (+, -) det
root(Ws, T) :- grep(T, null, Ws, T).

% op/4 (?, +, +, -) det <-- Root may or may not be fully ground
op(Root, null,                _X, Root).
op(Root, node([],      L, _R), X, T) :- op(Root, L, X, T).
op(Root, node([V|_Vs], L,  R), X, T) :-
    (X = V -> T = R ; op(Root, L, X, T)).

% next/3 (+, +, -) det
next(_X, null,               null).
next(_X, node([], L, R),     node([], L, R)).
next( X, node([V|Vs], L, R), T) :- ( X = V -> T = L ; T = node([V|Vs], L, R) ).

% grep/4 (+, +, +, -) det
grep(_Root, L, [],     node([], L, null)).
grep( Root, L, [V|Vs], node([V|Vs], L1, R)) :-
    next(V, L, L1), op(Root, L, V, T), grep(Root, T, Vs, R).

% ok/1 (+) det
ok(node([],  _L, _R)).

%% Driver

% matches_aux/5 (+, +, +, +, -) det
matches_aux(_Root, N, T, [],     Ns) :- ( ok(T) -> Ns = [N] ; Ns = [] ).
matches_aux( Root, N, T, [X|Xs], Ns) :-
    N1 is N + 1, op(Root, T, X, T1),
    ( ok(T) -> ( Ns = [N|Ns1], matches_aux(Root, N1, T1, Xs, Ns1) )
             ; matches_aux(Root, N1, T1, Xs, Ns) ).

% matches/3 (+, +, -) det
matches(Ws, Txt, Ns) :- root(Ws, Root), matches_aux(Root, 0, Root, Txt, Ns).

% :- root([1,2,1], Root).
% :- root([1,2,1,1,2], Root).
% :- matches([1,2,3,1,2], [1,2,1,2,3,1,2,3,1,2], Ns).
